Adding featured images to blog posts / adding Go blog post (#647)
cyberark/secretless-broker: docs/_posts/2019-03-04-go.md

---
layout: post
title: "Getting GOing"
date: 2019-03-04 09:00:00 -0600
author: Sigal Sax
categories: blog
published: true
image: go-gopher-illustration.jpg
thumb: go-gopher-illustration.jpg
image-alt: Go Gopher illustration

> excerpt: "For the past year, we on the Cyberark Conjur team have been working on our latest open source project in Go..."

Reviewer: > Why we chose Go for our latest open source project, the Secretless Broker and how this decision has impacted our development process

> excerpt: "For the past year, we on the Cyberark Conjur team have been working on our latest open source project in Go..."

Reviewer: ```suggestion
excerpt: "Why we chose Go for the Secretless Broker, and how this decision has impacted our development process."
```

---

From the very beginning, the original Golang developers had a clear goal - retain the positive attributes of the common programming languages while building one that was readable, simple, and high performing. Go’s light-weight model for fast compilation, extensive toolset, and native capabilities are what have propelled it toward worldwide adoption and success.  With these modern features and being backed by Google, it is no wonder why open source projects such as Docker, Kubernetes, and some in Cyberark, are being built in Go.  

## Get GOing with Secretless  

> For the past year, we on the Cyberark Conjur team have been working on our latest open source project in Go, the Secretless Broker. The Secretless Broker acts as a transparent proxy that injects the necessary credentials to target services directly, relieving client applications of the responsibility of handling sensitive data. Essentially, you can open a local connection to a credentialed resource without ever having to pass credentials to do so; all with Secretless. By removing the direct interaction between applications and target resources, the potential for a secret leak decreases significantly. 

Reviewer: ```suggestion
Over the past year, we on the Cyberark Conjur team have been working on our latest open source project in Go, the Secretless Broker. The Secretless Broker acts as a transparent proxy that injects the necessary credentials to target services directly, relieving client applications of the responsibility of handling sensitive data. Essentially, you can open a local connection to a credentialed resource without ever having to pass credentials to do so; all with Secretless. By removing the direct interaction between applications and target resources, the potential for a secret leak decreases significantly. 
```

> I was new to this project and especially new to the Go community, so naturally, I was curious why we made the design decision to build the Secretless Broker in Go. After working on features, participating in design discussions and debates during team standups, and having just attended the Go conference, it became clear why. We needed a system-level language that was as fast as the C-family but not as low level. Go gave us the perks of a low-level language while providing us with a toolset and native capabilities of a higher-level one.  

Reviewer: ```suggestion
I was new to this project and especially new to the Go community, so naturally, I was curious why we made the design decision to build the Secretless Broker in Go. After working on features, participating in design discussions and debates during team stand-ups, and having just attended the inaugural [GopherCon Israel](https://www.gophercon.org.il/), it became clear why. We needed a system-level language that was as fast as the C-family but not as low level. Go gave us the perks of a low-level language while providing us with a toolset and native capabilities of a higher-level one.
```

It also became difficult to consistently format our code, especially as our codebase increased in size and complexity. To fix this common problem, we added the `golint` automated tool into our workflow. Code linters run source code against an applied set of stylistic and syntax rules, raising flags when this structure is broken.  Linting has been an indispensable productivity tool for writing code and has even sped up our code review process.  

Documentation is essential to growing and sustaining the community around our project. We add to and tweak our README and contributor guides to encourage end-user self-efficiency and success. This has proven to be a time-consuming undertaking because our codebase is constantly evolving and being refactored. To our benefit, Go offers the `godoc` documentation tool to dynamically create documentation. This tool parses the code, relevant comments, and structure to generate documentation that evolves as fast as our code does. 

## … and beyond the toolchain 

> With Go, we can produce statically linked binaries that fully contain our applications and run them on different platforms. This capability supported our decision to move to Docker-based development for our building and testing.  

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

> Secretless is consumed as a Linux binary that is packaged inside a Docker image. That way, customers can easily get up and running once they grab our image. We also use Docker for our CI pipeline to build and publish our site, check styling and links, and run tests. We host these processes in Docker to ensure they receive the proper Go versioning for testing and to isolate their environments to ensure they always run in the same (and expected) way. 

Reviewer: ```suggestion
In the early stages of our project, we have been focusing on providing a native and secure Secretless solution for Kubernetes and Openshift environments. To accomplish this, we wanted to deliver Secretless so that it could be deployed as a sidecar container. Secretless is consumed as a Linux binary that is packaged inside a Docker image. That way, customers can easily get up and running once they grab our image. We also use Docker for our CI pipeline to build and publish our site, check styling and links, and run tests. We host these processes in Docker to ensure they receive the proper Go versioning for testing and to isolate their environments to ensure they always run in the same (and expected) way. 
```

Reviewer: @sigalsax please take a look at this suggestion

Reviewer 2: in the last sentence, "to ensure" is used twice. 
We can replace the 2nd "to ensure" with "so that"
OR rework the sentence (I sort of prefer this one. Otherwise, it is just a long sentence):
"We host these processes in Docker to ensure they receive the proper Go versioning for testing and to isolate their environments for process reproducibility."
What say you?

> When it comes to testing, I have mixed feelings about Docker. On one hand, without it, we cannot properly test our code and on the other, Docker has added an extra layer of complexity that makes our project difficult to understand and unapproachable. We find Docker beneficial because the alternative is a large assortment of messy Bash scripts and boilerplate code. Furthermore, contributors no longer need to fuss with tooling in order to configure their environment correctly - all they need is Docker and they are ready to get rolling. But with this assumption, we have added an extra pre-requisite that ultimately limits those who can contribute. In terms of testing norms, Docker-based development has forced us to break the standard Go testing conventions. Our tests have become custom to our project, wrapping the known Go tests with Docker-specific code. Consequently, it may become initially unclear to contributors how our tests work.  

Reviewer: ```suggestion
When it comes to testing, I have mixed feelings about Docker. Docker has added an extra layer of complexity and an additional prerequisite for contributing that may make our project less approachable. In terms of testing norms, Docker-based development has forced us to break the standard Go testing conventions. Our tests have become custom to our project, wrapping the known Go tests with Docker-specific code. Consequently, it may become initially unclear to contributors how our tests work.  However, in using Docker we are able to run integration tests that mimic a more realistic Secretless deployment, with Secretless running in a container and communicating with a database service, for example. Since Docker provides pre-built images and ensures complete process isolation, it makes the setup of test environments a breeze. We find Docker beneficial because the alternative is a large assortment of messy Bash scripts and boilerplate code. Furthermore, contributors no longer need to fuss with tooling in order to configure their environment correctly - all they need is Docker and they are ready to get rolling.
```

> When it comes to testing, I have mixed feelings about Docker. Docker has added an extra layer of complexity and an additional prerequisite for contributing that may make our project less approachable. In terms of testing norms, Docker-based development has forced us to break the standard Go testing conventions. Our tests have become custom to our project, wrapping the known Go tests with Docker-specific code. Consequently, it may become initially unclear to contributors how our tests work.  However, in using Docker we are able to run integration tests that mimic a more realistic Secretless deployment, with Secretless running in a container and communicating with a database service, for example. Since Docker provides pre-built images and ensures complete process isolation, it makes the setup of test environments a breeze. We find Docker beneficial because the alternative is a large assortment of messy Bash scripts and boilerplate code. Furthermore, contributors no longer need to fuss with tooling in order to configure their environment correctly - all they need is Docker and they are ready to get rolling.

Reviewer 2: ```suggestion
When it comes to testing, I have mixed feelings about Docker. Docker has added an extra layer of complexity and an additional prerequisite for contributing that may make our project less approachable. In terms of testing norms, Docker-based development has forced us to break the standard Go testing conventions. Our tests have become custom to our project, wrapping the known Go tests with Docker-specific code. Consequently, it may become initially unclear to contributors how our tests work.  However, in using Docker, we are able to run integration tests that mimic a more realistic Secretless deployment, with Secretless running in a container and communicating with a database service, for example. Since Docker provides pre-built images and ensures complete process isolation, it makes the setup of test environments a breeze. We find Docker beneficial because the alternative is a large assortment of messy Bash scripts and boilerplate code. Furthermore, contributors no longer need to fuss with tooling in order to configure their environment correctly - all they need is Docker and they are ready to get rolling.
```

Reviewer 2: I added a comma after "in using Docker"

> To curb these complexities and ensure a positive contributor experience, we are working to retain the integrity of Go testing through simplification and trying to provide better documentation. We hope this will eventually give our contributors the support and guidance they need to get involved and continue to contribute to our project.  

Reviewer: ```suggestion
To balance these complexities and ensure a positive contributor experience, we are working to enable standard local development while still leveraging the benefits of Docker when appropriate, and trying to provide better documentation so it's easy to get started. We hope this will eventually give our contributors the support and guidance they need to get involved and continue to contribute to our project.
```
